Add `editor.delete_directory()`

astrochili opened the pull request:
Fixes #8472

### Technical changes
This changeset adds a new function to editor scripts that allows the user to delete an existing directory. The function expects a resource path, i.e. a /-prefixed path from the root of the project, for example:

```lua
editor.delete_directory("/assets/generated")
```

astrochili committed:
Add editor.delete_directory()

https://github.com/defold/defold/issues/8472

astrochili commented:
I created it by looking at the PR [#8434](https://github.com/defold/defold/pull/8434)

I haven't figured out how to build the editor locally yet, but the code seems simple, please correct me if something is wrong 🙈.

astrochili commented:
https://github.com/defold/doc/pull/383

britzl commented:
> Haven't figured out how to build the editor locally yet, but the code seems simple. Please correct me if something is wrong 🙈.

I think it's good if you try to set up the editor so that you can build it. Submitting PRs "in the blind" so to speak is not a good practice :-) What kind of problems did you have when trying to set everything up?

AGulev commented:
I've just tested it and it doesn't work:
```
ERROR:EXT: hook "on_target_terminated" in /hooks.editor_script failed:
ERROR:EXT: /hooks.editor_script:5 class sun.nio.fs.UnixPath cannot be cast to class java.io.File (sun.nio.fs.UnixPath and java.io.File are in module java.base of loader 'bootstrap')
```

astrochili commented:
Sorry for your time, yes, that was brave 🫣. Gone to build the engine...

AGulev commented:
> Sorry for your time, yes, that was brave 🫣. Gone to build the editor...

We have #source-code channel in Discord. Feel free to ask your questions there. We will help you to build the engine (the first step) and the editor.

matgis commented:
No need to build the engine if you're making a change that only affects the editor in isolation. The `lein init archived-stable` command will download the latest stable engine and Bob binaries for you so you can skip setting up the whole engine build environment. Just make sure to prefix your branch name with `DEFEDIT-` and target the `editor-dev` branch with your PR. If your PR is accepted, it will be merged into the next editor update, so no need to wait for the next stable Defold release.

More info about the `lein init archived-stable` command here:
https://github.com/defold/defold/blob/dev/editor/README_BUILD.md#build-using-existing-engine-and-tools

AGulev commented:
@matgis oh, I didn't know that. Nice!

JCash commented:
Thanks for the PR!
I don't mean to pile on here, but perhaps @matgis could direct you in how to add a small test as well?

matgis commented:
@JCash I would feel more justified in asking for a test if our code they based this on had tests. 😅 Personally I think we can add tests later.

astrochili commented:
> More info about the `lein init archived-stable` command here:
> https://github.com/defold/defold/blob/dev/editor/README_BUILD.md#build-using-existing-engine-and-tools

Thanks. I built the editor and fixed the code. Updated commit.

Still not sure if the exceptions I added will work, or if any other checks are needed. But now directory deletion works 👍

astrochili commented:
**UPDATE:**
- Removed `NoSuchFileException` because it doesn't sense when deleting.
- Fixed `NotDirectoryException`, now it throws an exception to avoid deleting the file. Because creating and deleting files still available with `io.write` and `os.remove` as I understand.

Now there are only issues of code reuse in `editor-extensions.clj` and consistency of function call chains inside `fs.clj`. But  I'm not really good at this, first time writing in clojure.

matgis committed:
Prevent editor.delete_directory() from deleting certain directories

matgis commented:
@astrochili I added a commit with some additional protections against deleting anything below `.git` and `.internal`, as well as preventing you from deleting the project directory itself.

If it looks good to you, perhaps you could cherry-pick it into your PR?

https://github.com/defold/defold/commit/f18dec07aceea0c350c544115e82e30b656ed799

astrochili commented:
@matgis Added cherry-picked commit, looks good?

matgis reviewed (approved):
Looks good! I'm going to merge it to `editor-dev`. Once CI completes, you should get it in the next editor update.

vlaaad commented on editor/src/clj/editor/editor_extensions.clj:
```diff
@@ -466,20 +466,59 @@
         nil
         (catch FileAlreadyExistsException e
-          (throw (LuaError. (str "File already exists: " (.getMessage e)))))
+          (throw (LuaError. (str "Directory already exists: " (.getMessage e)))))
```
I'm fairly certain the error message should say "file already exists" and not "directory already exists". Directory already existing is not an error and should not throw exception. What should throw exception is if you wanted a directory at a certain path, and it turned out at that path there is a file instead, that's what the error it's about.

astrochili replied:
Oh really? I was thinking it's just a typo. So how to fix it back, a new [PR](https://github.com/defold/defold/pull/8481)?
